### PR TITLE
fix panic while deleting CSI plugins for missing job

### LINKED ISSUE #7757: 0.11 panic deregistering job
### Nomad version
0.11

### Issue

Our nomad servers are crashing while applying a delete job request from the raft log.

Based on our logs, a request to deregister+purge a job failed with a 500, though I can’t find any more details why. Within a few seconds we began seeing errors like this on each server

```
nomad.fsm: DeleteJob failed: error=“job not found”
nomad.fsm: deregistering job failed: error=“job not found”
```

Followed by panics:
```
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: panic: runtime error: invalid memory address or nil pointer dereference
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x17f190b]
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: goroutine 103 [running]:
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad/state.(*StateStore).deleteJobFromPlugin(0xc000614120, 0x48beee, 0xc00ef8aa40, 0xc0028f30e0, 0xc010d2da40, 0x0)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/state/state_store.go:1192 +0x37b
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad/state.(*StateStore).DeleteJobTxn(0xc000614120, 0x48beee, 0xc0035c26a0, 0x7, 0xc0035c2680, 0x8, 0xc00ef8aa40, 0x38061e0, 0xc010d2cfa0)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/state/state_store.go:1502 +0xafb
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad.(*nomadFSM).handleJobDeregister(0xc000326ee0, 0x48beee, 0xc0035c2680, 0x8, 0xc0035c26a0, 0x7, 0x1, 0xc00ef8aa40, 0xc0081e7a38, 0x17dbfc1)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/fsm.go:618 +0x1f6
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad.(*nomadFSM).applyDeregisterJob.func1(0xc00ef8aa40, 0x3142c01, 0xc00ef8aa40)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/fsm.go:565 +0x78
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad/state.(*StateStore).WithWriteTransaction(0xc000614120, 0xc0081e7b10, 0x0, 0x0)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/state/state_store.go:4873 +0x7b
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad.(*nomadFSM).applyDeregisterJob(0xc000326ee0, 0xc0040805a1, 0x4d, 0x4d, 0x48beee, 0x0, 0x0)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/fsm.go:564 +0x1f2
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/nomad.(*nomadFSM).Apply(0xc000326ee0, 0xc016b426e0, 0x577acc0, 0xbf9ea4efd4c4751b)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/nomad/fsm.go:208 +0x42d
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc016bd4120)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/fsm.go:90 +0x2c1
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).runFSM.func2(0xc000250a00, 0x40, 0x40)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/fsm.go:113 +0x75
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*Raft).runFSM(0xc000322900)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/fsm.go:219 +0x42f
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc000322900, 0xc003ff2c30)
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/state.go:146 +0x55
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]: created by github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft.(*raftState).goFunc
Apr 17 21:14:39 nomad-do-nyc1-61d4 nomad[19436]:         github.com/hashicorp/nomad/vendor/github.com/hashicorp/raft/state.go:144 +0x66
```

Based on the stack trace this [line (and appropriate comment above)](https://github.com/hashicorp/nomad/blob/master/nomad/state/state_store.go#L1192) was the culprit. We're not using any CSI plugins and this job had no other plugin configurations.

We were able to bring our servers back online with a patch that checked for nil task groups before ranging.

### Nomad Server logs (if appropriate)

I can share logs before and after the patch if you need.

### PR DESCRIPTION
**Pull request by michaeldwan**
This is what we deployed to production to fix the panic described in #7757.

### CODE REVIEW

File: nomad/state/state_store.go [hunk 1]
```diff
@@ -1187,15 +1187,14 @@ func (s *StateStore) deleteJobFromPlugin(index uint64, txn *memdb.Txn, job *stru
 	plugins := map[string]*structs.CSIPlugin{}
 
 	for _, a := range allocs {
-		tg := job.LookupTaskGroup(a.TaskGroup)
```
Comment (tgross) [inline nomad/state/state_store.go:1190]: Hi @michaeldwan I think the error we made here is that this line should have read:

```go
tg := a.job.LookupTaskGroup(a.TaskGroup)
```

It seemed strange to me that a job can not have the taskgroup for one of its own allocs, but my colleague @notnoop has pointed out to me that the alloc can have a view of the job that's stale relative to the job we've pulled from the state store, and that's why we're running into this panic.

Comment (michaeldwan) [reply nomad/state/state_store.go:1190]: That's roughly what we think happened too. We were actually working on a network issue that impacted consul but nomad seemed healthy through it.

I can make that change here.

Comment (tgross) [reply nomad/state/state_store.go:1190]: Gah, I typoed my example. It's `a.Job` (capital letter) 😊

Comment (michaeldwan) [reply nomad/state/state_store.go:1190]: I just made the change. There's a few other places LookupTaskGroup is being called but they appear to get both the job and alloc from the same snapshot and should be okay.

Comment (michaeldwan) [reply nomad/state/state_store.go:1190]: I made the same typo and pushed before waiting for tests in vagrant to run :)

**Pull request merged**